### PR TITLE
Minor noise estimation fixes for BGV and BFV

### PR DESCRIPTION
**Pull request by yspolyakov**
Refines BGV and BFV key switching estimates to correspond to Appendix B of https://eprint.iacr.org/2021/204

### CODE REVIEW

File: src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp [hunk 1]
```diff
@@ -130,8 +130,10 @@ uint32_t FindLevelsToDrop(usint multiplicativeDepth, std::shared_ptr<CryptoParam
     auto noiseKS = [&](uint32_t n, double logqPrev, double w) -> double {
         if (scalTechnique == HYBRID)
             return k * (numPartQ * delta(n) * Berr + delta(n) * Bkey + 1.0) / 2;
-        else
-            return delta(n) * (floor(logqPrev / (log(2) * dcrtBits)) + 1) * w * Berr;
+        else {
+            double numDigitsPerTower = relinWindow == 0 ? 1 : floor(dcrtBits / relinWindow) + 1;
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp:134]: a few parentheses would improve the readability. also std::floor doesn't make any sense here as this is a division of 2 unsigned integers (even one of them declared as "int").

File: src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp [hunk 2]
```diff
@@ -121,15 +121,17 @@ bool ParameterGenerationBFVRNS::ParamsGenBFVRNS(std::shared_ptr<CryptoParameters
...
-            // of moduli at this point and use an upper bound for numDigits
-            return (multiplicativeDepth + 1) * delta(n) * Berr;
-        else
-            return delta(n) * (floor(logqPrev / (std::log(2) * dcrtBits)) + 1) * w * Berr;
+            // of digits and moduli at this point and use upper bounds
+            double numTowers = ceil(static_cast<double>(logqPrev) / dcrtBits);
+            return numTowers * (delta(n) * Berr + delta(n) * Bkey + 1.0) / 2.0;
+        }
+        else {
+            double numDigitsPerTower = digitSize == 0 ? 1 : floor(dcrtBits / digitSize) + 1;
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp:132]: a few parentheses would improve the readability. also std::floor doesn't make any sense here as this is a division of 2 unsigned integers (even one of them declared as "int").

File: src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp [hunk 3]
```diff
@@ -111,14 +111,14 @@ BGVNoiseEstimates ParameterGenerationBGVRNS::computeNoiseEstimates(
         }
         int relinBase       = pow(2.0, digitSize);
         int modSizeEstimate = DCRT_MODULUS::MAX_SIZE;
-        int numWindows      = floor(modSizeEstimate / log(relinBase)) + 1;
+        int numWindows      = floor(modSizeEstimate / digitSize) + 1;
```
Comment (dsuponitskiy) [inline src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp:114]: std::floor doesn't make any sense here as this is a division of 2 unsigned integers (even one of them declared as "int").

File: src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp (hunk 1 above)
Comment (yspolyakov) [reply src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp:134]: Ok with parentheses. Please removed floor as it is not needed (it is already implied as you said).

File: src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp (hunk 2 above)
Comment (yspolyakov) [reply src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp:132]: Same as above.

File: src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp (hunk 3 above)
Comment (yspolyakov) [reply src/pke/lib/scheme/bgvrns/bgvrns-parametergeneration.cpp:114]: Agreed

**Review by dsuponitskiy — APPROVED** [review]
Looks good

File: src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp [hunk 4]
```diff
@@ -130,8 +130,10 @@ uint32_t FindLevelsToDrop(usint multiplicativeDepth, std::shared_ptr<CryptoParam
     auto noiseKS = [&](uint32_t n, double logqPrev, double w) -> double {
         if (scalTechnique == HYBRID)
             return k * (numPartQ * delta(n) * Berr + delta(n) * Bkey + 1.0) / 2;
-        else
-            return delta(n) * (floor(logqPrev / (log(2) * dcrtBits)) + 1) * w * Berr;
+        else {
+            double numDigitsPerTower = (relinWindow == 0) ? 1 : ((dcrtBits / relinWindow) + 1);
+            return delta(n) * numDigitsPerTower * (floor(logqPrev / (log(2) * dcrtBits)) + 1) * w * Berr;
```
Comment (andreea-alexandru) [inline src/pke/lib/scheme/bfvrns/bfvrns-leveledshe.cpp:135]: Is there a 1/2 factor missing?

File: src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp [hunk 5]
```diff
@@ -121,15 +121,17 @@ bool ParameterGenerationBFVRNS::ParamsGenBFVRNS(std::shared_ptr<CryptoParameters
...
+        if (ksTech == HYBRID) {
             // conservative estimate for HYBRID to avoid the use of method of
             // iterative approximations; we do not know the number
-            // of moduli at this point and use an upper bound for numDigits
-            return (multiplicativeDepth + 1) * delta(n) * Berr;
-        else
-            return delta(n) * (floor(logqPrev / (std::log(2) * dcrtBits)) + 1) * w * Berr;
+            // of digits and moduli at this point and use upper bounds
+            double numTowers = ceil(static_cast<double>(logqPrev) / dcrtBits);
+            return numTowers * (delta(n) * Berr + delta(n) * Bkey + 1.0) / 2.0;
```
Comment (andreea-alexandru) [inline src/pke/lib/scheme/bfvrns/bfvrns-parametergeneration.cpp:129]: Could this be instead (numTowers * delta(n) * Berr + delta(n) * Bkey + 1.0) / 2.0?

**Pull request merged**